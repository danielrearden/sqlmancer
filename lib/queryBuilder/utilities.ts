import Knex from 'knex'
import { QueryBuilderContext, Dialect } from '../types'

export function getAlias(name: string, context: QueryBuilderContext): string {
  const index = name.substring(0, 1).toLowerCase()
  if (!context.alias[index]) {
    context.alias[index] = 1
  } else {
    context.alias[index] = context.alias[index]! + 1
  }
  return `${index}${context.alias[index]}`
}

export function getDirection(enumValue: string): string {
  const enumValues: Record<string, string> = {
    ASC: 'asc',
    DESC: 'desc',
  }

  if (!enumValues[enumValue]) {
    throw new Error(`Invalid direction value: ${JSON.stringify(enumValue)}`)
  }

  return enumValues[enumValue]
}

export function getAggregateFunction(key: string): string {
  const aggregates: Record<string, string> = {
    avg: 'avg',
    max: 'max',
    min: 'min',
    sum: 'sum',
  }

  if (!aggregates[key]) {
    throw new Error(`Invalid aggregate function name: ${JSON.stringify(key)}`)
  }

  return aggregates[key]
}

export function getComparisonExpression(
  knex: Knex,
  dialect: Dialect,
  column: string | Knex.Ref<any, any>,
  operator: string,
  value: any
): Knex.Raw {
  if (value === null && (operator === 'equal' || operator === 'notEqual')) {
    return knex.raw(`${column} is ${operator === 'notEqual' ? 'not ' : ''}null`)
  }

  switch (operator) {
    case 'equal':
      return knex.raw(`${column} = ?`, [value])
    case 'notEqual':
      return knex.raw(`${column} <> ?`, [value])
    case 'in':
      return knex.raw(`${column} in (${value.map(() => '?').join(', ')})`, value)
    case 'notIn':
      return knex.raw(`${column} not in (${value.map(() => '?').join(', ')})`, value)
    case 'greaterThan':
      return knex.raw(`${column} > ?`, [value])
    case 'greaterThanOrEqual':
      return knex.raw(`${column} >= ?`, [value])
    case 'lessThan':
      return knex.raw(`${column} < ?`, [value])
    case 'lessThanOrEqual':
      return knex.raw(`${column} <= ?`, [value])
    case 'like':
      return knex.raw(`${column} like ?`, [value])
    case 'notLike':
      return knex.raw(`${column} not like ?`, [value])
  }

  if (dialect === 'postgres') {
    switch (operator) {
      case 'contains':
        return knex.raw(`${column} @> ?`, [value])
      case 'containedBy':
        return knex.raw(`${column} <@ ?`, [value])
      case 'overlaps':
        return knex.raw(`${column} && ?`, [value])
      case 'iLike':
        return knex.raw(`${column} ilike ?`, [value])
      case 'notILike':
        return knex.raw(`${column} not ilike ?`, [value])
      case 'hasKey':
        return knex.raw(`${column} \\? ?`, [value])
      case 'hasAnyKeys':
        return knex.raw(`${column} \\?| ?::text[]`, [value])
      case 'hasAllKeys':
        return knex.raw(`${column} \\?& ?::text[]`, [value])
    }
  }

  if (dialect === 'mysql' || dialect === 'mariadb') {
    switch (operator) {
      case 'contains':
        return knex.raw(`json_contains(${column}, ?)`, [value])
      case 'containedBy':
        return knex.raw(`json_contains(?, ${column})`, [value])
      case 'hasKey':
        return knex.raw(`json_contains_path(${column}, 'all', concat('$.', ?))`, [value])
      case 'hasAnyKeys':
        return knex.raw(`json_contains_path(${column}, 'one', ${value.map(() => `concat('$.', ?)`).join(', ')})`, value)
      case 'hasAllKeys':
        return knex.raw(`json_contains_path(${column}, 'all', ${value.map(() => `concat('$.', ?)`).join(', ')})`, value)
    }
  }

  throw new Error(`Unsupported operator "${operator}" for dialect ${dialect}`)
}

export function getJsonObjectFunctionByDialect(dialect: Dialect) {
  switch (dialect) {
    case 'postgres':
      return 'json_build_object'
    case 'mysql':
      return 'json_object'
    case 'mariadb':
      return 'json_object'
    case 'sqlite':
      return 'json_object'
  }
}

export function getJsonAggregateExpressionByDialect(
  dialect: Dialect,
  isArray: boolean,
  expression: Knex.Ref<any, any>
): string {
  switch (dialect) {
    case 'postgres':
      return isArray
        ? `coalesce(nullif(json_agg(${expression})::text, '[null]'), '[]')::json`
        : `json_agg(${expression}) -> 0`
    case 'mysql':
      return isArray
        ? `if(json_arrayagg(${expression}) is null, json_array(), json_arrayagg(${expression}))`
        : `json_extract(json_arrayagg(${expression}), '$[0]')`
    case 'mariadb':
      return isArray
        ? `if(json_arrayagg(${expression}) is null, json_array(), json_arrayagg(${expression}))`
        : `json_extract(json_arrayagg(${expression}), '$[0]')`
    case 'sqlite':
      return isArray
        ? `json_group_array(json(${expression}))`
        : `json_extract(json_group_array(json(${expression})), '$[0]')`
  }
}
