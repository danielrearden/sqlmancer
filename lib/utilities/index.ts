export * from './makeNullable'
export * from './parseResolveInfo'
export * from './unwrap'
