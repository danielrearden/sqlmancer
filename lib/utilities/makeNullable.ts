import { GraphQLType, GraphQLNonNull } from 'graphql'

export function makeNullable(type: GraphQLType) {
  if (type instanceof GraphQLNonNull) {
    return type.ofType
  }
  return type as any
}
