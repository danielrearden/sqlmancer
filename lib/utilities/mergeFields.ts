import { GraphQLField } from 'graphql'

export function mergeFields(fields: GraphQLField<any, any>[]) {
  return fields.reduce((acc, field) => {
    const args = acc.args
    field.args.forEach((arg) => {
      if (!args.find(({ name }) => name === arg.name)) {
        args.push(arg)
      }
    })
    return { ...acc }
  })
}
