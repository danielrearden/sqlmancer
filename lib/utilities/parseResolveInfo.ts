import { GraphQLResolveInfo } from 'graphql'
import { parse, FieldsByTypeName, ResolveTree } from 'graphql-parse-resolve-info'

export interface FlattenedResolveTree {
  name: string
  alias: string
  args: Record<string, any>
  fields: Record<string, FlattenedResolveTree>
}

export function parseResolveInfo(info: GraphQLResolveInfo): FlattenedResolveTree | null {
  const parsedInfo = parse(info)
  if (!parsedInfo) {
    return null
  }

  const { name, alias, args, fieldsByTypeName } = parsedInfo as ResolveTree
  return {
    name,
    alias,
    args,
    fields: flattenFieldsByType(fieldsByTypeName),
  }
}

function flattenFieldsByType(fieldsByType: FieldsByTypeName) {
  return Object.keys(fieldsByType).reduce((acc, typeName) => {
    Object.keys(fieldsByType[typeName]).forEach((fieldName) => {
      const { name, alias, args, fieldsByTypeName } = fieldsByType[typeName][fieldName]
      acc[fieldName] = { name, alias, args, fields: flattenFieldsByType(fieldsByTypeName) }
    })
    return acc
  }, {} as Record<string, FlattenedResolveTree>)
}
