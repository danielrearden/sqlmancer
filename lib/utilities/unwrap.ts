import { GraphQLType, GraphQLNonNull, GraphQLList, GraphQLNamedType } from 'graphql'

export function unwrap(type: GraphQLType) {
  let wrappedType = type

  while (wrappedType instanceof GraphQLList || wrappedType instanceof GraphQLNonNull) {
    wrappedType = wrappedType.ofType
  }

  return wrappedType as GraphQLNamedType
}
