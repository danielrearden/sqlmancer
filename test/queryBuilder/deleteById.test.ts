import { withDialects } from '../utilities'

describe('DeleteBuilder', () => {
  withDialects((client, rollback) => {
    describe('basic queries', () => {
      test('no additional options', async () => {
        const builder = client.models.Actor.deleteById(20)
        const { sql, bindings } = builder.toQueryBuilder().toSQL()
        await rollback(builder, (result) => expect(result).toBeBoolean())
        expect(sql).toMatchSnapshot()
        expect(bindings).toMatchSnapshot()
      })
    })
  })
})
