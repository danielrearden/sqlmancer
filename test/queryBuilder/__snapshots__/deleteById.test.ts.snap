// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteBuilder mysql basic queries no additional options 1`] = `"delete from \`actor\` where \`actor\`.\`actor_id\` = ?"`;

exports[`DeleteBuilder mysql basic queries no additional options 2`] = `
Array [
  20,
]
`;

exports[`DeleteBuilder postgres basic queries no additional options 1`] = `"delete from \\"actor\\" where \\"actor\\".\\"actor_id\\" = ?"`;

exports[`DeleteBuilder postgres basic queries no additional options 2`] = `
Array [
  20,
]
`;

exports[`DeleteBuilder sqlite basic queries no additional options 1`] = `"delete from \`actor\` where \`actor\`.\`actor_id\` = ?"`;

exports[`DeleteBuilder sqlite basic queries no additional options 2`] = `
Array [
  20,
]
`;
